mw_hmac_test.go

```diff
@@ -69,10 +73,48 @@ func getHMACAuthChain(spec *APISpec) http.Handler {
 	return chain
 }
 
+type testAuthFailEventHandler struct {
+	cb func(config.EventMessage)
+}
+
+func (w *testAuthFailEventHandler) Init(handlerConf interface{}) error {
+	return nil
+}
+
+func (w *testAuthFailEventHandler) HandleEvent(em config.EventMessage) {
+	w.cb(em)
+}
+
+func waitTimeout(wg *sync.WaitGroup, timeout time.Duration) bool {
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		wg.Wait()
+	}()
+	select {
+	case <-c:
+		return false // completed normally
+	case <-time.After(timeout):
+		return true // timed out
+	}
+}
+
 func TestHMACAuthSessionPass(t *testing.T) {
 	spec := createSpecTest(t, hmacAuthDef)
 	session := createHMACAuthSession()
 
+	// Should not receive an AuthFailure event
+	var eventWG sync.WaitGroup
+	var eventAuthFailState bool = false
+	eventWG.Add(1)
+	cb := func(em config.EventMessage) {
+		eventAuthFailState = true
+		eventWG.Done()
+	}
```

Comment:
You could simplify the logic here by having `waitTimeout` return whether the waitgroup finished or timed out. In the former case, it's the same as if `eventAuthFailState` was set to true.

So you can get rid of the bool, and use `if !waitTimedOut(&eventWG, 20*time.Millisecond) { ...` below.